peps/pep-0750.rst

```diff
@@ -53,837 +33,1089 @@ deferred pending further experience with f-strings. Work on this PEP was
 
-    name = "World"
-    greeting = greet"Hello {name} nice to meet you"
-    assert greeting == "Hello WORLD nice to meet you!"
+The ``Template`` Type
+---------------------
 
-Tag strings extract more than just a callable from the ``Interpolation``. They also
-provide Python string formatting info, as well as the original text:
+Template strings evaluate to an instance of a new type, ``Template``, found
+in the ``<TBC>`` module (proposed: ``types``):
 
 .. code-block:: python
 
-    def greet(*args: Decoded | Interpolation) -> str:
-        """Interpolations can have string formatting specs and conversions."""
-        result = []
-        for arg in args:
-            match arg:
-                case Decoded() as decoded:
-                    result.append(decoded)
-                case getvalue, raw, conversion, format_spec:  # Unpack
-                    gv = f"gv: {getvalue()}"
-                    r = f"r: {raw}"
-                    c = f"c: {conversion}"
-                    f = f"f: {format_spec}"
-                    result.append(", ".join([gv, r, c, f]))
+    @dataclass(frozen=True)
+    class Template:
+        args: Sequence[str | Interpolation]
 
-        return f"{''.join(result)}!"
+The ``args`` attribute provides access to the string parts and
+any interpolations in the literal:
+
+.. code-block:: python
 
     name = "World"
-    assert greet"Hello {name!r:s}" == "Hello gv: World, r: name, c: r, f: s!"
+    template = t"Hello {name}"
+    assert isinstance(template.args[0], str)
+    assert isinstance(template.args[1], Interpolation)
+    assert template.args[0] == "Hello "
+    assert template.args[1].value == "World"
 
-You can see each of the ``Interpolation`` parts getting extracted:
+The use of ``@dataclass`` in the definition of ``Template`` above (and 
+``Interpolation`` below) is meant to be suggestive; the exact implementation 
+in CPython may differ, but developers can expect that ``Template`` instances 
+can be constructed and utilized in the same way as a typical dataclass.
 
-- The lambda expression to call and get the value in the scope it was defined
-- The raw string of the interpolation (``name``)
-- The Python "conversion" field (``r``)
-- Any `format specification <https://docs.python.org/3/library/string.html#format-specification-mini-language>`_
-  (``s``)
+See `Interleaving of Template.args`_ below for more information on how the 
+``args`` attribute is structured.
 
-Specification
-=============
 
-In the rest of this specification, ``my_tag`` will be used for an arbitrary tag.
-For example:
+The ``Interpolation`` Type
+--------------------------
 
-.. code-block:: python
+The ``Interpolation`` type represents an expression inside a template string. 
+Like ``Template``, it is a new concrete type found in the ``types`` module:
 
-    def mytag(*args):
-        return args
+.. code-block:: python
 
-    trade = 'shrubberies'
-    mytag'Did you say "{trade}"?'
+    @dataclass(frozen=True)
```

Comment:
As above, will this be a dataclass at runtime (`assert dataclasses.is_dataclass(...) is True`)?